### DISCUSSION TITLE
How to properly setup Renovate for React Native?

### DISCUSSION THREAD

**Discussion by yangm97** [renovatebot/renovate#11122] [Request Help]
Disclaimer: I've been running Renovate for a while in a react native project and it's an awesome tool. However it seems like I'm not using it to its full potential and got bitten more often than not, clearly due the complexity of a react native project and my own lack of attention to the Renovate configuration. Given that, I've been trying to collect the required pieces to get a "less invasive" and working renovate config for a react native project.

So React Native is one hell of a beast when it comes to setting up Renovate. The biggest pain point comes from the very fact react native projects need to be set up from a boilerplate so upgrading is not as straightforward as bumping versions in `package.json`. Another issue which comes with these projects is the need to match package versions, sometimes across package managers, in order to get consistent with a given react native release.

So far I could only think about the nuclear option of making renovate ignore all dependencies listed on [RnDiffApp](https://github.com/react-native-community/rn-diff-purge/tree/release/0.64.2/RnDiffApp) in order to ensure version integrity, at cost of upgrade visibility. Unless we could mark a set of dependencies to only be upgraded given a custom command, in this case we would essentially shift the responsibility of upgrading these dependencies to `react-native` cli but I'm not aware of such facility in Renovate.

So, while that addresses "first party" dependencies we still need to sort some "third party" dependencies (aka the ones added by yourself). Matching peer dependencies appears to be a bigger concern for react native projects than it is for average ones, especially given react native itself doesn't seem to adhere to SemVer and will introduce breaking changes even on patch releases, so we don't want to upgrade "third party" dependencies to become potentially incompatible, especially given tests (both automated and human) could let some edge case issues be introduced because of this mismatch. This could be addressed by #1864.

Finally, there's also a need to update the `Podfile.lock` when running nodejs renovations. I've tried to replicate #10335 but it didn't work for me (see bellow).

Without further-ado I will post my current preset:

```
{
  "extends": [
    "config:base",
    ":switchToGradleLite"
  ],
  "node": {
    "skipInstalls": false,
    "postUpgradeTasks": {
      "commands": [
        "cd ios && bundle install && bundle exec pod install"
      ],
      "fileFilters": [
        "ios/Podfile.lock"
      ],
      "executionMode": "update"
    }
  },
  "packageRules": [
    {
      "groupName": "react native monorepo",
      "matchSourceUrlPrefixes": [
        "https://github.com/facebook/react",
        "https://github.com/facebook/react-native"
      ],
      "matchPackageNames": [
        "react-native-macos",
        "react-native-web",
        "react-native-windows"
      ]
    },
    {
      "matchPackageNames": [
        "com.facebook.react:react-native"
      ],
      "enabled": false
    }
  ]
}
```

Notes:
- I'm running [self-hosted](https://gitlab.com/myio-os/renovate)
- Unexpected updates still slip in (and some will get automerged so beware!), haven't applied the nuclear strategy yet

**Comment by HonkingGoose** [comment]
> * Some monorepos aren't working with this config (i.e. `ts-jest` is not being grouped within the jest monorepo)

Try adding `group:jestPlusTSJest` to your `extends` array 😉 

https://docs.renovatebot.com/presets-group/#groupjestplustsjest

I'll let the more qualified maintainers help you with the other stuff. 😄

**Reply by yangm97** [reply]
@HonkingGoose but it looks like `group:recommended` already extends `group:jestPlusTSJest` and `config:base` already extends `group:recommended`... couldn't it be a bug?

**Reply by HonkingGoose** [reply]
Whoops! 😄 

You're right, `group:recommended` includes `group:jestPlusTSJest` and the `config:base` preset includes `group:recommended`.  Renovate _should_ group `ts-jest` with the Jest monorepo.

@rarkins and @viceice do you have any ideas what might be going wrong here?

**Reply by viceice** [reply]
It only groups for major updates
![image](https://user-images.githubusercontent.com/1798109/128481143-d20f751b-7f6c-4461-8f13-bde3ad690b08.png)

**Reply by HonkingGoose** [reply]
`ts-jest` only matches Jest versioning for major versions, so that's the only time you should group `jest` and `ts-jest` together.

Quote from [`kulshekhar/ts-jest/README.md`](https://github.com/kulshekhar/ts-jest/blob/master/README.md)

> ## Versioning
> 
> We **DO NOT** use [SemVer](https://semver.org/) for versioning. Though you can think about SemVer when reading our version, except our major number follows the one of Jest. For the versions available, see the [tags on this repository](https://github.com/kulshekhar/ts-jest/tags).

**Reply by yangm97** [reply]
Thanks for the clarification, I've updated the OP accordingly.

**Comment by yangm97** [comment]
Ok so now I'm getting an error under some PRs

> ##### File name: package.json
>
> ```
> Command failed: cd ios && bundle install && bundle exec pod install
> /bin/sh: 1: bundle: not found
> ```
>

So I pulled `renovate/renovate` locally and:

```sh
$ which bundle
/usr/local/ruby/3.0.0/bin/bundle
```

What is happening?

**Reply by rarkins** [reply]
@yangm97 I assume you're running self-hosted? Are you using a "full" image and not slim?

@viceice any ideas on this?

**Reply by yangm97** [reply]
Yes, self hosted: https://gitlab.com/myio-os/renovate 

Temporarily commented out the allowed commands to get rid of this error from MRs, can reenable whenever you guys want to debug.

**Reply by rarkins** [reply]
@viceice is that running slim or full by default?

If full, then maybe our post update commands child process isn't passing the Path through correctly?

**Reply by viceice** [reply]
He's using the slim version, at it's the dind template.

BTW: autodiscover without filter is a very bad idea. Someone can easily invite your bot into it's own repo and steal your account via pipeline https://docs.renovatebot.com/gitlab-bot-security/.


post upgrade tasks are run in main renovate image, so the slim image is missing ruby and bundler here. So for that specific use case i would recomend to use the full version if that works.


In future this can be easily done by: https://github.com/containerbase/buildpack/issues/101

**Reply by yangm97** [reply]
🤦‍♂️ can't believe I didn't notice the slim tag after looking at the merged yaml so many times...

@viceice I've set up the env var for the filter but gonna double check just in case, thanks.
